### ISSUE TITLE
Support of unnamed parameters in Microsoft.Data.Sqlite

### ISSUE THREAD

**Issue by TheDevKnight** [dotnet/efcore#24480]
Hi,

is the support of unnamed parameters in Microsoft.Data.Sqlite possible? 
I migrate an old codebase of sqlite code where we use lots of unnamed parameters in our queries. 

Like
`this.CreateCommand("ATTACH DATABASE ? AS ? ", filePath, schemaName)`

the only solution that i found was to change my queries to

`this.CreateCommand("ATTACH DATABASE @p1 AS @p2 ", filePath, schemaName)`

and modify the code who fill the parameters in the SqliteCommand.

Are there any plan to add support for unnamed parameters.

Regards

**Comment by bricelam** [comment]
See also https://github.com/aspnet/Microsoft.Data.Sqlite/issues/8

**Comment by roji** [comment]
Related note (as discussed in triage): PostgreSQL's native parameter format is positional ($1, $2, $3). Npgsql parses the SQL and rewrites named parameter placeholders (@p, @q) with the native parameters as part of query processing. I've always hated this, and once the ADO.NET batching API goes in I also plan to implement a "raw" mode where the user just uses positional parameters, and SQL no longer gets parsed.

At that point the driver would support both named (with client-side parsing/replacing) and positional - but not with the same command obviously. I was thinking of doing this via a connection string and/or via an NpgsqlCommandBehavior or similar, but nothing finalized yet.

**Comment by MortenBoysen** [comment]
This is a feature me and my team could also use. We are porting a codebase from System.Data.SQLite to Microsoft.Data.Sqlite and the code is using positional parameters. We have created an ugly hack that allows our code to still use positional parameters in the SQL statements, but it is not ideal. Support for this in the same manner as System.Data.SQLite would be very welcome.

**Comment by roji** [comment]
@bricelam I've been thinking about this lately, and as written above I'll most probably implement support for this in Npgsql for 6.0. I will probably simply check if the first parameter has its ParameterName set. If so, I'll assume the rest do as well and go into the named parameter flow as today (where the driver parses and replaces placeholders in the SQL). If not, I'll assume positional parameters are being used, and send them directly without needing to parse/replace.

There are unfortunately some additional complications (Npgsql also parses and replaces to detect semicolons, for batching).

**Comment by TheDevKnight** [comment]
@roji this sounds good to me 👍

**Comment by lukasf** [comment]
Please add unnames parameters support. I was pretty surprised that this does not work, given the fact that it is a very common way of parameter handling for Sqlite. I also have legacy code that uses unnamed parameters a lot. I was hoping that moving to Microsoft.Data.Sqlite was easy, but this lack of an important feature makes it rather difficult.

**Comment by bricelam** [comment]
Everyone, please remember to upvote (👍) the issue description. It's really the only way we can weigh the importance of this issue relative to the 1.7k others in this repo.

**Comment by lukasf** [comment]
Done. The surprising thing is that the API already suggests that this is supported. Property names are nullable strings, and you can access and set properties either by name or by index.

Binding parameters by index on sqlite side is even easier than binding them by name, since you don't need to lookup the index from name. You even save one step of work when going by index. Creating a PR for this should be more or less trivial.

**Comment by bricelam** [comment]
The complexity comes when you mix named and positional parameters.

**Comment by lukasf** [comment]
I don't think that mixed names and positional parameters are a real use case. You either use named or unnamed parameter in a query. We could just disallow mixing both and avoid the complexity.

**Comment by roji** [comment]
> The surprising thing is that the API already suggests that this is supported. Property names are nullable strings, and you can access and set properties either by name or by index. Only when you actually use these APIs, you will find out that it does not work.

This is a result of how the ADO.NET (System.Data) abstraction is designed; it must support both databases with positional and named parameters. For example, it's possible to not set the parameter name on SQL Server, but then one gets a runtime error. This in itself isn't a good reason to implement both techniques on any specific database driver.

However, if System.Data.Sqlite does support positional placeholders and supporting them makes it easier for users to migrate, that may be a good reason to do this.

**Comment by bricelam** [comment]
> We could just disallow mixing both and avoid the complexity.

Yes, I'd be fine with an implementation that threw in this case.

**Comment by bricelam** [comment]
If anyone is interested in contributing this, we'd recommend proposing a design here before submitting a PR. For example how would you detect the case when ordered and named parameters are being mixed, and how would cases like those listed in https://github.com/aspnet/Microsoft.Data.Sqlite/issues/8#issue-29480451 be handled. How would you handle error cases like too many or not enough parameters? Also, how would ordered parameters work with batched (multi-statement) commands?
